extensions/walletProviders/utxo/src/extension.ts

```diff
@@ -0,0 +1,356 @@
+    })
+
+    if ('fromAddress' in payload) {
+      privateKeysToSign = await this.privateKeysFromAddress(tx, payload)
+    } else if ('fromUTXO' in payload) {
+      privateKeysToSign = await this.privateKeysFromUTXO(tx, payload)
+    }
+
+    new Set(privateKeysToSign).forEach((key) => {
+      tx.sign(new PrivateKey(key))
+    })
+
+    const rawTransaction = tx.serialize(this.config?.skipAllChecks)
+
+    const response = await this.utxoRpc.sendRawTransaction(rawTransaction)
+
+    if (!response?.result) {
+      throw new Error(JSON.stringify(response.error))
+    }
+
+    return response.result
+  }
+
+  supportedNetworks: Network[] = [
+    Network.BITCOIN,
+    Network.DOGECOIN,
+    Network.LITECOIN,
+    Network.BITCOIN_TESTNET,
+    Network.DOGECOIN_TESTNET,
+    Network.LITECOIN_TESTNET,
+  ]
+
+  private async privateKeysFromAddress(
+    transaction: BtcBasedTransaction,
+    body: BtcTransactionFromAddress,
+  ): Promise<Array<string>> {
+    if (body.fromAddress.length === 0 && !this.config?.skipAllChecks) {
+      throw new Error('No fromAddress provided')
+    }
+
+    let totalInputs = 0
+    let totalOutputs = body.fee ? toSatoshis(body.fee) : 0
+    for (const item of transaction.outputs) {
+      totalOutputs += item.satoshis
+    }
+    const privateKeysToSign: string[] = []
+
+    for (const item of body.fromAddress) {
+      if (totalInputs >= totalOutputs) {
+        break
+      }
+
+      const utxos = await this.getUtxos(item, totalOutputs, totalInputs)
+      for (const utxo of utxos) {
+        totalInputs += utxo.value
+        transaction.from([
+          Transaction.UnspentOutput.fromObject({
+            txId: utxo.txHash,
+            outputIndex: utxo.index,
+            script: Script.fromAddress(utxo.address).toString(),
+            satoshis: toSatoshis(utxo.value),
+          }),
+        ])
+
+        privateKeysToSign.push(item.privateKey)
+      }
+    }
+
+    return privateKeysToSign
+  }
+
+  private async getUtxos(item: BtcTransactionFromAddressSource, totalOutputs: number, totalInputs: number) {
+    return this.connector.get<UtxoResponse>({
+      path: `/v3/data/utxos?chain=${this.sdkConfig.network}&address=${
+          item.address
+      }&totalValue=${fromSatoshis(totalOutputs - totalInputs)}`,
+    })
+  }
+
+  private async privateKeysFromUTXO(
+    transaction: Transaction,
+    body: BtcTransactionFromUTXO,
+  ): Promise<Array<string>> {
+    if (body.fromUTXO.length === 0 && !this.config?.skipAllChecks) {
+      throw new Error('No fromUTXO provided')
+    }
+
+    const privateKeysToSign: string[] = []
+    const utxos: BtcUTXO[] = []
+
+    const filteredUtxos = await this.getUtxoBatch(body)
+
+    for (let i = 0; i < filteredUtxos.length; i++) {
+      const utxo = filteredUtxos[i]
+      if (utxo === null || !utxo.address) continue
+
+      const utxoItem = body.fromUTXO[i]
+
+      utxos.push(utxo)
+
+      transaction.from([
+        Transaction.UnspentOutput.fromObject({
```

Comment:
if this is common code, please check if it works for doge, I remember there were some troubles with tx preparaion for doge, it prepares differently from other btc-based

Reply:
you're right, fixed